File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/persistence/DatabaseCartRepositoryTest.java
package com.tarbadev.witchcraft.persistence;

import com.tarbadev.witchcraft.domain.entity.Cart;
import com.tarbadev.witchcraft.domain.entity.Item;
import com.tarbadev.witchcraft.domain.entity.Recipe;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Java6Assertions.assertThat;


@RunWith(SpringRunner.class)
@DataJpaTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
@ActiveProfiles("test")
public class DatabaseCartRepositoryTest {
  @Autowired private TestEntityManager entityManager;
  @Autowired private CartEntityRepository cartEntityRepository;

  private DatabaseCartRepository subject;

  @Before
  public void setUp() {
    subject = new DatabaseCartRepository(cartEntityRepository);
  }

  @Test
  public void findAll() {
    entityManager.persist(CartEntity.builder().build());
    entityManager.persist(CartEntity.builder().build());
    entityManager.persist(CartEntity.builder().build());

    entityManager.flush();
    entityManager.clear();

    assertThat(subject.findAll().size()).isEqualTo(3);
  }

  @Test
  public void save() {
    List<Recipe> recipes = Collections.singletonList(
        EntityToDomain.recipeMapper(
            entityManager.persistAndFlush(
                RecipeEntity.builder()
                    .name("Lasagna")
                    .ingredients(Arrays.asList(
                        IngredientEntity.builder()
                            .name("Ingredient 3")
                            .unit("cup")
                            .quantity(2.0)
                            .build(),
                        IngredientEntity.builder()
                            .name("Ingredient 1")
                            .unit("lb")
                            .quantity(2.0)
                            .build(),
                        IngredientEntity.builder()
                            .name("Ingredient 2")
                            .unit("oz")
                            .quantity(8.0)
                            .build()
                    ))
                    .steps(Collections.emptyList())
                    .build()
            )
        )
    );

    entityManager.clear();

    List<Item> items = Arrays.asList(
        Item.builder()
            .name("Ingredient 1")
            .unit("lb")
            .quantity(2.0)
            .build(),
        Item.builder()
            .name("Ingredient 2")
            .unit("oz")
            .quantity(8.0)
            .build(),
        Item.builder()
            .name("Ingredient 3")
            .unit("cup")
            .quantity(2.0)
            .build()
    );
    Cart cart = subject.save(Cart.builder()
        .recipes(recipes)
        .items(items)
        .build());

    entityManager.clear();

    CartEntity savedCart = entityManager.find(CartEntity.class, cart.getId());

    assertThat(savedCart.getItems().size()).isEqualTo(3);
    assertThat(savedCart.getRecipes().size()).isEqualTo(1);
  }

  @Test
  public void save_createsDateForCart() {
    CartEntity cart = CartEntity.builder().build();
    assertThat(cart.getCreatedAt()).isNull();

    entityManager.persistAndFlush(cart);

    assertThat(cart.getCreatedAt()).isNotNull();
  }

  @Test
  public void findById() {
    CartEntity cart = entityManager.persistFlushFind(
        CartEntity.builder()
            .build()
    );

    entityManager.clear();

    assertThat(subject.findById(cart.getId())).isEqualToComparingFieldByFieldRecursively(cart);
  }

  @Test
  public void findById_returnsItemsOrderedAlphabetically() {
    CartEntity cart = entityManager.persistAndFlush(
        CartEntity.builder()
            .items(
                Arrays.asList(
                    ItemEntity.builder()
                        .name("Parsley")
                        .build(),
                    ItemEntity.builder()
                        .name("Tomatoes")
                        .build(),
                    ItemEntity.builder()
                        .name("Beef")
                        .build()
                )
            )
            .build()
    );

    entityManager.clear();

    List<Item> items = cart.getItems().stream()
        .map(itemEntity -> Item.builder()
            .id(itemEntity.getId())
            .name(itemEntity.getName())
            .quantity(itemEntity.getQuantity())
            .unit(itemEntity.getUnit())
            .build())
        .sorted(Comparator.comparing(Item::getName))
        .collect(Collectors.toList());

    List<Item> savedItems = subject.findById(cart.getId()).getItems();
    assertThat(savedItems).isEqualTo(items);
    assertThat(savedItems.get(0)).isEqualTo(items.get(0));
    assertThat(savedItems.get(1)).isEqualTo(items.get(1));
    assertThat(savedItems.get(2)).isEqualTo(items.get(2));
  }
}

File: WitchcraftApp/src/test/kotlin/com/tarbadev/witchcraft/carts/persistence/DatabaseCartRepositoryTest.kt
package com.tarbadev.witchcraft.carts.persistence

import com.tarbadev.witchcraft.carts.domain.entity.Cart
import com.tarbadev.witchcraft.carts.domain.entity.Item
import com.tarbadev.witchcraft.carts.persistence.entity.CartEntity
import com.tarbadev.witchcraft.carts.persistence.entity.ItemEntity
import com.tarbadev.witchcraft.carts.persistence.repository.CartEntityRepository
import com.tarbadev.witchcraft.carts.persistence.repository.DatabaseCartRepository
import com.tarbadev.witchcraft.recipes.persistence.entity.IngredientEntity
import com.tarbadev.witchcraft.recipes.persistence.entity.RecipeEntity
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.util.Arrays.asList

@ExtendWith(SpringExtension::class)
@DataJpaTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
class DatabaseCartRepositoryTest(
    @Autowired private val entityManager: TestEntityManager,
    @Autowired private val cartEntityRepository: CartEntityRepository
) {
    private var databaseCartRepository: DatabaseCartRepository = DatabaseCartRepository(cartEntityRepository)

    @BeforeEach
    fun setup() {
        cartEntityRepository.deleteAll()
    }

    @Test
    fun findAll() {
        entityManager.persist(CartEntity())
        entityManager.persist(CartEntity())
        entityManager.persist(CartEntity())

        entityManager.flush()
        entityManager.clear()

        assertEquals(3, databaseCartRepository.findAll().size)
    }

    @Test
    fun save() {
        val recipes = listOf(entityManager.persistAndFlush(
            RecipeEntity(
                name = "Lasagna",
                ingredients = asList(
                    IngredientEntity(
                        name = "Ingredient 3",
                        unit = "cup",
                        quantity = 2.0
                    ),
                    IngredientEntity(
                        name = "Ingredient 1",
                        unit = "lb",
                        quantity = 2.0
                    ),
                    IngredientEntity(
                        name = "Ingredient 2",
                        unit = "oz",
                        quantity = 8.0
                    )
                ),
                steps = emptySet()
            )
        )
            .toRecipe())

        entityManager.clear()

        val items = asList(
            Item(
                name = "Ingredient 1",
                unit = "lb",
                quantity = 2.0,
                enabled = true
            ),
            Item(
                name = "Ingredient 2",
                unit = "oz",
                quantity = 8.0,
                enabled = true
            ),
            Item(
                name = "Ingredient 3",
                unit = "cup",
                quantity = 2.0,
                enabled = true
            )
        )
        val (id) = databaseCartRepository.save(Cart(
            recipes = recipes,
            items = items
        ))

        entityManager.clear()

        val savedCart = entityManager.find(CartEntity::class.java, id)

        assertEquals(3, savedCart.items.size)
        assertEquals(1, savedCart.recipes.size)
    }

    @Test
    fun findById() {
        val cart = entityManager.persistAndFlush(CartEntity()).toCart()

        entityManager.clear()

        assertEquals(cart, databaseCartRepository.findById(cart.id))
    }

    @Test
    fun findById_returnsItemsOrderedAlphabetically() {
        val cart = entityManager.persistAndFlush(
            CartEntity(
                items = listOf(
                    ItemEntity(name = "Parsley"),
                    ItemEntity(name = "Onions"),
                    ItemEntity(name = "Beef")
                )
            )
        )

        entityManager.clear()

        val items = cart.items
            .map { itemEntity -> itemEntity.toItem() }
            .sortedBy { it.name }

        assertIterableEquals(items, databaseCartRepository.findById(cart.id)!!.items)
    }

    @Test
    fun findById_returnsNullIfNotFound() {
        assertNull(databaseCartRepository.findById(1))
    }
}